R/content-tools.R

```diff
@@ -31,20 +45,39 @@ on_load({
       }
 
       maybe_echo_tool(result, echo = echo)
+      maybe_on_tool_result(result, on_tool_result)
       yield(result)
     }
   })
 
   # invoke_tools_async is intentionally *not* an _async_ generator, instead it
   # is a generator that returns promises. This lets the caller decide if the
   # tasks should be run in parallel or sequentially.
-  invoke_tools_async <- coro::generator(function(turn, tools, echo = "none") {
+  invoke_tools_async <- coro::generator(function(
+    turn,
+    tools,
+    echo = "none",
+    on_tool_request = NULL,
+    on_tool_result = NULL
```

Comment:
```suggestion
    on_tool_request = function(...) invisible(),
    on_tool_result = function(...) invisible()
```

I think that makes it more clear what the typically input will be and I think allows you to simplify some of the `maybe_` helpers.